src/server/game/Instances/InstanceSaveMgr.cpp

```diff
@@ -286,8 +286,38 @@ void InstanceSaveManager::LoadInstances()
     CharacterDatabase.DirectExecute("UPDATE corpse SET instanceId = 0 WHERE instanceId > 0 AND instanceId NOT IN (SELECT id FROM instance)");
     CharacterDatabase.DirectExecute("UPDATE characters AS tmp LEFT JOIN instance ON tmp.instance_id = instance.id SET tmp.instance_id = 0 WHERE tmp.instance_id > 0 AND instance.id IS NULL");
 
+    // Reduce instance ID numbers when they climb too high, to prevent overflow
+    QueryResult result = CharacterDatabase.Query("SELECT MIN(id), MAX(id) FROM instance");
+    if (result)
+    {
+        uint32 minId = (*result)[0].GetUInt32();
+        uint32 maxId = (*result)[1].GetUInt32();
+
+        // If IDs have reached 4 billion, re-number
+        if (maxId >= uint32(0xEE6B2800))
+        {
+            TC_LOG_INFO("server.loading", "Instance IDs exceed 4000000000, re-numbering...");
+
+            uint32 gap = minId - 1;
+            SQLTransaction trans = CharacterDatabase.BeginTransaction();
+
+            trans->PAppend("UPDATE instance SET id = id-%u ORDER BY id ASC", gap);
```

Comment:
this will fail if someone has a very low id and keeps extending the save forever.

By renumbering I meant also compacting the ids, for example
[4, 23, 546, 2132] would become [1, 2, 3, 4], not, [1, 20, 543, 2129]